core/psh/psh.c

```diff
@@ -972,7 +972,181 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_login(void)
+{
+	int insz = -1, shdwlines = 0, code;
+	const int maxinsz = 16;
+	char c, indata[maxinsz+1], indatasave[maxinsz+1];
+	char shdw[30+1];
+	struct termios mode;
+
+	tcgetattr(0, &mode);
+    mode.c_lflag &= ~(ECHO | ICANON);
```

Comment:
You should definately use canonical mode with echo on - and disable echo just for password input.

That way You don't need to parse special chars by hand (like backspace).

(please note that I've already written that in the previous PR, please let me know why You decided not to follow my advice)

Reply:
Yes, that was my mistake. Using canonical mode shaves a lot of code from getting credentials loops. It comes out that tty, at the moment of psh_login run, already is in canonical mode. I should probably force the canonical mode just in case. In current implementation right after login the psh itself setups the tty modes so it probably doesn`t matter what state the tty was before. 

Of course: unsetting the ICANON flag was removed. ICANON flag is forced to be present at the begining of the psh_login.

Reply:
I've already wrote that in the previous PR (https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/35#issuecomment-810953550) - because of that was annoyed and surprised to see another implementation with parsing special chars by hand.

I believe the POSIX standard ensures the console should be in canonical mode at the beginning of every executed program.